qpage.man
'\"macro stdmacro
.nr X
.TH qpage 1 "05/08/99" "Thomas Dwyer III"
.\" Copyright (c) 1995-1999 Thomas Dwyer III
.SH NAME
qpage \- SNPP client/server for sending messages to an alpha-numeric pager
.SH SYNOPSIS
.B qpage
[
.B \-dhimQv
] [
.BI \-a " [+][dd+]hhmm
|
.BI \-a " YYMMDDHHMMSS
] [
.BI \-c " coverage
] [
.BI \-f " from
] [
.BI \-l " level
] [
.BI \-p " pagerid
] [
.BI \-P " pagerid
] [
.BI \-s " server
] [
.I message
]
.LP
.B qpage
[
.B \-d
] [
.BI \-C " config
]
.BI \-q " interval
.SH DESCRIPTION
.B QuickPage
sends messages to a paging terminal using the SNPP and IXO (also known
as TAP) protocols.  It is normally used with no options other than
a recipient and the message text, in which case the message is
sent to the SNPP server where it is submitted to a page queue
to be sent by a separate daemon process.
Page groups and duty schedules are supported.
Status notification messages
indicating the success or failure of a page
are sent via e-mail to submitters of
high-priority (level 0) pages.
.LP
If no message is specified on the command line,
the user is prompted for the message text.  Characters are read from
standard input until a line containing only a period is
received, or until EOF is received.  All occurrences of whitespace
within the message text are reduced to a single space character.
.LP
Messages received by the
.B QuickPage
daemon that are longer than
the maximum page length (see below) are split into multiple segments
and sent as separate pages.
Each message segment is prefixed with its segment number.
The maximum number of segments a into which a message will be split
is configurable.
If a message exceeds this limit, the last allowed segment will be
prefixed with 't', indicating that the message is truncated;
otherwise, the last segment is prefixed with 'e' to indicate the
end of the message.
Messages that fit into a single page are not prefixed with segment numbers.
All message segments are prefixed with CALLerid information (see RFC-1861),
if available.
.LP
.SH OPTIONS
The following options are supported: 
.TP
.BI \-a
Send the page at the specified time.  A time specification prefixed
with a plus sign (+) is added to the current time.  The
.B QuickPage
daemon will not send any pages with a timestamp newer than the current
time.
This feature may be used to provide a simple alarm clock function
for meeting reminders, etc.
This option applies only to the pagerid(s) specified by the next
.B \-p
option.
If the specified time is in the format YYMMDDHHMMSS, the two-digit
year is interpreted as follows: if the last two digits of the specified
year are in the range 00-49 and the last two digits of the current
year are in the range 50-99, assume the specified time is in the next
century.  Otherwise assume the specified time is in the current
century.
.TP
.BI \-c
Use a different coverage area or paging service.  This option is
only useful if the recipient has more than one pager and/or more
than one paging service. 
This option applies only to the pagerid(s) specified by the next
.B \-p
option.
.TP
.BI \-C
Specify an alternate configuration file.  This option may only
be used on the
.B QuickPage
server.
.TP
.BI \-d
Debug mode.  Very verbose and probably not interesting for the
normal user.  This option is provided for debugging purposes only.
.TP
.BI \-f
Specify who the page is from.  This option specifies the argument to
the SNPP
.B CALLerid
command.  Every message segment will be
prefixed with the value specified by this option unless disabled
by the msgprefix keyword in the configuration file.  A null
argument (
.B \-f
"") may be specified to suppress the message prefix for a particular
message.  However, this will also suppress e-mail status notification.
The default is the userid of the person running
.B QuickPage.
.TP
.BI \-h
Help.  Print a brief summary of the command line options.
.TP
.BI \-i
Use interactive mode.  The page is sent immediately and summary
messages are printed during the course of the transaction.
This option may only be used on a machine with a physically
attached modem.  This option is intended for debugging purposes
only and should not be used in a production environment.  The
.B \-d
option, when used in conjunction with this option, can be
very effective in troubleshooting problems between the SNPP
server and the remote paging terminal.
.TP
.BI \-l
Specify the service level for this page.  The service level must be
a number between 0 and 11, inclusive.
The TME protocol specifies service level as follows:
.LP
.RS
.RS
.PD 0
0 - Priority
.TP
1 - Normal (default)
.TP
2 - Five minutes
.TP
3 - Fifteen minutes
.TP
4 - One hour
.TP
5 - Four hours
.TP
6 - Twelve hours
.TP
7 - Twenty Four hours
.TP
8 - Carrier specific '1'
.TP
9 - Carrier specific '2'
.TP
10 - Carrier specific '3'
.TP
11 - Carrier specific '4'
.PD
.RE
.LP
With the exception of level zero, service levels have no
meaning to
.B QuickPage
but they are accepted for compatibility with other programs.
Any service level specified by the user is passed on to
the SNPP server.
For pages submitted with a service level of zero, the
.B QuickPage
daemon will send an e-mail message to
the submitter notifying them whether the page succeeded (i.e. it
was successfully transmitted to the paging service) or failed.
This option applies only to the pagerid(s) specified by the next
.B \-p
option.
.RE
.TP
.BI \-m
This option tells
.B QuickPage
to read an e-mail message from standard input.  A page is
constructed by concatenating the From: header (XXX), the
Subject: header (YYY), and lines from the message
body (ZZZ) as follows:
.LP
.RS
.RS
XXX (Subj: YYY) ZZZ ...
.RE
.LP
Minimal support is provided for multipart MIME messages.  The first
part with a Content-Type: of text/plain will be processed.  The
remaining parts will be discarded.  This will reduce the problems
associated with people accidentally mailing binaries to their pagers.
.LP
The X-sun-attachment type is not supported and should not be used.
MIME is a documented standard.
The X-sun-attachment type is neither documented nor standard.
However,
.B QuickPage
will still use the From: and Subject: lines (but no message body)
from such messages for those people who insist on using it anyway.
.LP
A line starting with two dashes (--) in the message body (or in
a MIME text part) is assumed to be the start of a signature and
will cause
.B QuickPage
to stop processing the e-mail message at that point.
.RE
.TP
.BI \-p
Specify the pagerid of the intended recipient.
Multiple recipients may be specified by separating
their pager IDs with commas (e.g. pagerid1,pagerid2,pagerid3).
No spaces are permitted between recipients and the comma
separator characters.
Any
.B \-a,
.B \-c,
or
.B \-l
options previously encountered on the command line are reset
to their default values after this option is processed.
This option may occur multiple times on the command line, each
possibly preceded by different
.B \-a,
.B \-c,
or
.B \-l
options.
.TP
.BI \-P
Like
.B \-p
but does not reset
.B \-a,
.B \-c,
or
.B \-l
options previously encountered on the command line.
.TP
.BI \-q
Start a
.B QuickPage
daemon.  This option causes
.B QuickPage
to detach itself from the controlling terminal and run as a daemon
process.  This option requires a numerical argument.  If the argument
is greater than zero, it specifies the number of seconds
.B QuickPage
should sleep between queue runs.  An argument less than zero signals
.B QuickPage
to accept incoming SNPP connections and write new pages to the page
queue, but to never process the page queue.  An argument of zero
signals
.B QuickPage
not to listen for incoming SNPP connections, but to process the
page queue one time and then exit.
See the NOTES section below for more information.
.TP
.BI \-Q
Print the current contents of the page queue.  This option may
only be used on the
.B QuickPage
server.  Unless invoked by root or DAEMONUSER, this option will
probably fail because of insufficient permissions.
.TP
.BI \-s
Specify the name(s) of the SNPP server(s).  The hostnames used by
a client to contact a server are
determined by checking the following in this order:
.LP
.RS
.RS
.PD 0
the
.B \-s
option
.TP
the environment variable SNPP_SERVER
.TP
the contents of the compiled in filename
.TP
the compiled in hostname
.PD
.RE
.LP
Multiple hostnames are permitted by separating them with commas.
If multiple servers are used, it is assumed that they all have identical
copies of the configuration file.
.RE
.TP
.BI \-v
Print the version of
.B QuickPage
and exit.
.LP
The
.B QuickPage
server requires a configuration file.  This file is read
one time during startup and again upon receipt of SIGHUP.
The configuration file
is made up of major and minor keyword=value pairs.
Major keywords must start at the beginning of a line.
Minor keywords must contain leading whitespace.
All keywords must be immediately followed with an equal sign (=).
Spaces are permitted between the equal sign and the value.  The
value may not contain whitespace.
Keywords may appear in any order with the following four exceptions:
.RS
.LP
Minor keywords must be grouped together under their respective major keyword.
.LP
Modems must be defined before any services that reference them.
.LP
Services must be defined before any pagers that reference them.
.LP
Pagers must be defined before any groups that reference them.
.RE
.LP
The major keywords are:
.RS
.TP 15
administrator
The e-mail address of the
.B QuickPage
administrator.  If defined, e-mail
notification will be sent to the
.B QuickPage
administrator whenever a page fails (i.e. when
.B maxtries
has been exceeded).
This address is also used in the Reply-To: header for status notification
messages sent to users who submit high-priority pages. 
.TP
forcehostname
(true, false, or @mailhost)
Force the destination
address to be qualified with a hostname
when sending e-mail status notification to users.
This option can be used when the
.B QuickPage
daemon is running on a machine that does not
handle unqualified e-mail addresses correctly.
If the value of this keyword starts with the '@' character,
it will be appended as-is to unqualified e-mail addresses.
If the value of this keyword is "true" then the submitter's hostname
will be appended to such addresses.
The default is false (do not append hostnames).
.TP
identtimeout
The number of seconds to wait for a reply
before giving up on RFC-1413 queries.
An integer less than or equal to zero disables ident functionality.
The default is 10 seconds.
.TP
include
If present, this keyword specifies the name of another configuration
file that should be processed at this point.  Processing of the
current file resumes after the specified file has been processed.
.B QuickPage
makes no attempt to prevent infinite recursion; do not use this keyword
in multiple files that point at each other.
.TP
pidfile
If present, this keyword specifies a file into which the server
should write its process ID.  The file is is not opened until
after root permissions have been dropped.  If DAEMONUSER cannot
open the file for writing (or if the file does not exist and
DAEMONUSER does not have permission to create it), this
keyword is silently ignored.
.TP
sigfile
If present, this keyword specifies a file containing an alternate
signature that
.B QuickPage
should append to e-mail status notification
messages sent to page submitters.  Use sigfile=/dev/null to suppress
the signature completely.
.TP
synchronous
(true or false)
Whether or not queue runs are synchronized with new pages.
If true, the submission of a new page initiates a queue run without
waiting for the normal sleep counter (set by the \-q option).  If false,
the page is queued and the server waits for the time remaining on
the sleep counter (if any) before starting another queue run.
The default is true.
.TP
lockdir
The location of the lock directory.
This keyword may be used to override the compiled in location
of the lock directory.
It should rarely be necessary to specify this keyword.
.TP
snpptimeout
The number of seconds to wait for an SNPP command before
terminating the connection.
The default is 60 seconds.
.TP
queuedir
The location of the page queue.
There is no default queue directory; this option must be specified
in the configuration file.
.TP
modem
The start of a modem definition.
The argument to this keyword is the name of a modem device (e.g. hayes).
This keyword has the following minor keywords:
.RS
.TP 15
text
Optional text specified by the administrator.  This text may not
contain whitespace.  However, underscores are converted to spaces
when the value of this keyword is read.
This keyword is not interpreted by
.B QuickPage
and is only provided for the administrator's convenience.
.TP
device
The name of the device the modem is physically connected
to, such as /dev/cua/a.
There is no default device; this option must be specified
in the configuration file.
.TP
initcmd
The initialization command for this modem.
The initialization command must cause the modem to respond
with OK.
The default initialization command is ATZ.
.TP
dialcmd
The dial command for this modem.  The dial command specified here
should not include a phone number.
The default dial command is ATDT.
.RE
.TP
service
The start of a paging service definition.
The argument to this keyword is the name of the paging service.
If a service named
.B default
exists, the values of its minor keywords are used as defaults for
all other service definitions.
This keyword has the following minor keywords:
.RS
.TP 15
text
Optional text specified by the administrator.  This text may not
contain whitespace.  However, underscores are converted to spaces
when the value of this keyword is read.
This keyword is not interpreted by
.B QuickPage
and is only provided for the administrator's convenience.
.TP
device
One or more names of modem definitions.  Multiple modems may be
specified by separating them with commas.  During a queue run,
.B QuickPage
will try each modem in the list until it finds one that can be
opened successfully (i.e. it is not already being used by a dialin
user or by some other process).
The remaining modems in the list will be ignored until the next queue
run when the list will be searched again.
In no event will
.B QuickPage
ever have more than one modem open at any given time; pages queued for
different services are not processed concurrently.
For backward compatibility with previous releases, a physical device
path may be specified here.  However, this will not be supported in
future releases.  Modem definitions should be used instead.
.TP
dialcmd
[obsolete] The command that the modem should use to connect to the remote
paging service.  This keyword is accepted for backward compatibility
and will not be supported in future releases.  If specified,
this keyword overrides the dialcmd keyword in all modem
specifications used by this service.
.TP
phone
The phone number of the paging service.  The specified phone number
will be appended to the modem's dialcmd when calling the paging
service.
.TP
password
The password to use when logging into the remote paging service.
The IXO specification defines the password as an optional six
character alphanumeric string.  Most paging services in the
United States do not require a password.
The default is an empty string ("").
.TP
baudrate
The speed to use while talking to the modem.
The default is 300 baud.
.TP
parity
The parity to use (even, odd, or none) while talking to the modem.
The default is even.
.TP
maxmsgsize
The maximum number of characters allowed in a single
page segment.  This size includes the message prefix
and 9 bytes of protocol overhead.
The default is 80 characters.
.TP
maxpages
The maximum number of page segments to send when a message exceeds
.B maxmsgsize.
The value of this option must be between 1 and 9, inclusive.
The default is 9 page segments.
.TP
maxtries
The number of times to attempt sending a page before giving up.
If the modem is busy (i.e. tip or cu is currently using it) or if
the modem returns BUSY in response to the dial command, the counter
is not incremented.  Thus a busy service will cause
.B QuickPage
to retry the page forever.
The default is 6.
.TP
identfrom
(true or false)
This keyword specifies whether to use the ident response as the
message prefix if no CALLerid SNPP command is received.
The default is true.
.TP
allowpid
(true or false)
This keyword specifies whether the
.B QuickPage
daemon will accept numeric pagerids for pagers not specified in
the configuration file.  The default service is
used for such pagerids unless the user explicitly chooses a
different service.
The default for this keyword is false.
.TP
msgprefix
(true or false)
Whether to prepend the sender's name (the CALLerid information)
to the beginning of each page segment.  This keyword should be
set to false for service definitions used for numeric pagers.
The default for this keyword is true.
.RE
.TP
pager
The start of a pager definition.
The argument to this keyword is the username associated with the pager.
This username will be specified by the
.B \-p
option on the command line when running
.B QuickPage
in client mode.
This keyword has the following minor keywords:
.RS
.TP 15
text
Optional text specified by the administrator.  This text may not
contain whitespace.  However, underscores are converted to spaces
when the value of this keyword is read.
Pager specifications containing this keyword will be listed by
the XWHO SNPP command.
See the NOTES section below for more information.
.TP
pagerid
The pagerid sent to the remote paging service for this pager.
.TP
service
The default service to use for this pager.
.RE
.TP
group
The start of a group definition.  The argument to this keyword is
the name of a new page group.  This keyword has the following minor
keywords:
.RS
.TP 15
text
Optional text specified by the administrator.  This text may not
contain whitespace.  However, underscores are converted to spaces
when the value of this keyword is read.
Group specifications containing this keyword will be listed by
the XWHO SNPP command.
See the NOTES section below for more information.
.TP
member
A member of this group.  The member must have a valid

CHANGES


		Changes to QuickPage from v3.2 to v3.3
		---------------------------------------

	- Building QuickPage has been greatly simplified by GNU
	  autoconf.  All you need to do is unpack the tar file,
	  type "./configure" and then type "make".  It's that simple.

	- QuickPage now has builtin support for Wietse Venema's
	  "tcp_wrappers" package.  See the INSTALL guide for more
	  information.

	- In the configuration file, groups and pagers are no longer
	  permitted to have the same names.  Every group and pager
	  entry must have a unique name.  This should help administrators
	  detect configuration errors more easily.

	- When invoking sendmail to sending status notifications, 
	  the end of the message is now signified by closing standard
	  input to the child instead of writing the <CR>.<CR> sequence.
	  This should allow QuickPage to work with programs other than
	  sendmail (such as the qmail).

	- Null characters received from the remote paging service
	  are now ignored.

	- QuickPage now gives up after 5 attempts when trying to log
	  into the paging service.  Previous versions used to retry
	  forever or until the remote modem disconnected.

	- An initial carriage return is sent, followed by a 1 second
	  delay, before sending the initialization string to the modem.
	  Without this, the initialization command sometimes failed on
	  some modems due to unclean disconnects from previous
	  connections.

	- Multiple modems are supported.  A new "modem" keyword
	  in the configuration file allows modem-specific
	  initialization and dialing commands to be associated with
	  each modem.  A paging service may include references to
	  multiple modem definitions.  This does not mean pages
	  will be processed in parallel--see the man page for details.

	- A new "synchronous" major keyword specifies whether new
	  pages should be processed immediately or whether the server
	  should wait until the normal sleep time (-q argument) as
	  expired.  The default is true (process new pages immediately).

	- A new "lockdir" major keyword can be used to override the
	  compiled in lock directory.

	- A new "pidfile" major keyword causes the server to write its
	  process ID to the specified file.

	- A new "sigfile" major keyword specifies an alternate signature
	  to be used in e-mail status notification messages.

	- A new "initcmd" minor keyword is allowed in the modem and
	  service specifications.

	- A new "phone" minor keyword is allowed in the service
	  specification.  If present, the value of this keyword
	  will be appended to the dialcmd when dialing the modem.

	- A new "msgprefix" minor keyword is allowed in the service
	  specification to indicate whether the sender's name is
	  prefixed to each message segment.

	- The "dialcmd" minor keyword used for service definitions is now
	  obsolete.  It is still supported in this release for backward
	  compatibility, but it will not be supported in future releases.
	  Instead, use the new "phone" keyword in combination with a
	  modem specification.

	- Better error checking is done when the configuration file is
	  being read.

	- A "default" paging service now always exists, even if it
	  is not specified in the configuration file.  Compiled in
	  default settings for the default service may be overridden
	  in the configuration file.

	- The configuration file now supports the "include" keyword
	  to specify the name of another configuration file that should
	  be inserted at that point.  This allows the administrator
	  to split the configuration file into multiple files if
	  desired.

	- The page queue is locked during queue runs so that only
	  one child can process the queue at a time.

	- Bad pages are renamed to Bxxxxxxx to make it easier for
	  administrators to locate and examine/remove them.

	- Lock files are now created with mode 666, regardless of
	  the umask in effect when QuickPage is started.

	- Previous versions read the configuration file many times
	  (for each SNPP connection, each queue run, etc.).  Now,
	  the configuration file is only read during startup and
	  on receipt of SIGHUP.

	- A new SNPP command, XWHO, can be used by clients to
	  determine which pagers are supported by the server.
	  This command is not part of the official SNPP protocol
	  as specified in RFC-1861 and it is only implemented here
	  to overcome the current protocol's deficiency.  If the
	  protocol is ever revised in the future to include this
	  functionality, support for the XWHO command will be
	  dropped in favor of whatever facilities are specified
	  at that time.  Do not consider XWHO a stable command.
	  It may go away in a future version of QuickPage without
	  notice.

	- The modem, service, pager, and group major keywords now
	  allow a "text" minor keyword that can be used to associate
	  an arbitrary text string with each of these entries.  Any
	  pager or group entry that contains a text string will be
	  shown by the SNPP "XWHO" command (not an official command
	  specified by RFC-1861).

	- Previous versions forked two daemon processes, one to handle
	  incoming SNPP connections and another one to process the page
	  queue.  Starting with v3.3, QuickPage only uses one main daemon
	  process which forks short-term child processes when necessary.

	- The return code of the client now correctly reflects whether
	  the page was successfully transmitted to the server.

	- Characters are stripped to 7 bits before transmitting them
	  to the paging service.  Also, control characters are escaped
	  by first sending 0x1A and then sending the ASCII character
	  obtained by adding 0x40 to the ASCII value of the control
	  character.  This is required by the IXO/TAP protocol.

	- When using interactive mode with the -i flag, page groups
	  are now expanded correctly.

	- QuickPage now prints an error message if the -C option is
	  specified on the client side.

	- Root privileges are always dropped, even during interactive

qpage.c
/*
** QuickPage
**
**	This program sends pages to an alpha-numeric pager.
**
**	Author:		Thomas Dwyer III <tomiii@qpage.org>
**	Version:	3.3
**
*/
#include	"qpage.h"


/*
** Global variables
*/
#ifndef lint
static char	sccsid[] = "@(#)qpage.c  3.36  07/26/98  tomiii@qpage.org";
#endif
char		*ConfigFile = NULL;
int		Debug = 0;
int		Interactive = FALSE;
int		Silent = FALSE;


/*
** do_version()
*/
void
do_version(void)
{
	printf("\n");
	printf("QuickPage v%s, Copyright 1995-98 by Thomas Dwyer III\n",
		VERSION);
	printf("\n");
}


/*
** do_usage()
*/
void
do_usage(char *prog)
{
	do_version();
	printf("Usage: %s [ options ] [ pagerid ] [ message ]\n", prog);
	printf("\n");
	printf("Options:\n");
	printf("    -a [+]hhmm    send the page at the specified time\n");
	printf("    -c coverage   coverage area (service name) for pager\n");
	printf("    -C config     use an alternate configuration file\n");
	printf("    -d            debug mode\n");
	printf("    -f from       who the page is from\n");
	printf("    -h            help (this screen)\n");
#ifndef CLIENT_ONLY
	printf("    -i            interactive mode\n");
#endif
	printf("    -l level      level of service\n");
	printf("    -m            read an e-mail message as client input\n");
	printf("    -p pager      pager id to recieve the message\n");
	printf("    -P pager      same as -p but don't reset -[acl]\n");
#ifndef CLIENT_ONLY
	printf("    -q xxx        process the queue every xxx seconds\n");
	printf("    -Q            show the pages currently in the queue\n");
#endif
	printf("    -s server     name of SNPP the server (default=%s)\n",
		SNPP_SERVER);
	printf("    -v            print qpage version and exit\n");
	printf("\n");
}


/*
** command_line_message()
**
** Build a message from the remaining command line arguments.  If
** there is no message on the command line, prompt the user to enter
** the message from standard input.
**
**	Input:
**		argc - the number of arguments
**		argv - an array of argument strings
**
**	Returns:
**		the message text
*/
char *
command_line_message(int argc, char **argv)
{
	char	*buf;
	char	*src;
	char	*dst;
	int	buflen;
	int	len;
	int	i;


	buf = NULL;
	dst = NULL;
	len = 0;
	buflen = 0;

	/*
	** Loop through each argument, appending it (followed by a space)
	** to the message buffer.
	*/
	for (i=0; i<argc; i++) {
		src = argv[i];
		len = strlen(src);

		/*
		** make sure the buffer is big enough to hold this argument
		*/
		while (buflen-(dst-buf) <= len + 1) {
			buf = (char *)my_realloc(buf, buflen+BUFCHUNKSIZE);
			buflen += BUFCHUNKSIZE;

			if (dst == NULL)
				buf[0] = '\0';

			dst = buf + strlen(buf);
		}

		(void)strcpy(dst, src);
		dst += len;
		*dst++ = ' ';
		*dst = '\0';
	}

	/*
	** nuke the trailing space
	*/
	if (dst)
		*(dst-1) = '\0';

	/*
	** check for an empty message
	*/
	if (buf && *buf == '\0') {
		free(buf);
		buf = NULL;
	}

	if (buf == NULL) {
		if (Silent == FALSE)
			printf("Enter the message; end with '.' on a "
				"line by itself\n");

		buf = getinput(stdin, FALSE);

		if (Silent == FALSE && buf && *buf)
			printf("Ok.\n");
	}

	if (buf && *buf)
		return(buf);
	else
		return(NULL);
}


/*
** add_recipients()
**
** This function adds recipients to the PAGE structure.
**
**	Input:
**		p - the PAGE structure
**		str - the string of recipients to parse
**		coverage - the coverage in effect
**		holduntil - the hold-time in effect
**		level - the service level in effect
**
**	Returns:
**		nothing
*/
static void
add_recipients(PAGE *p, char *str, char *coverage, time_t holduntil, int level)
{
	char		*ptr;
	rcpt_t		*tmp;


	if ((ptr = strtok(str, ",/:")) == NULL)
		return;

	do {
		tmp = (void *)malloc(sizeof(*tmp));
		(void)memset((char *)tmp, 0, sizeof(*tmp));
		tmp->next = p->rcpts;
		tmp->pager = strdup(ptr);
		tmp->holduntil = holduntil;
		tmp->coverage = coverage ? strdup(coverage) : NULL;
		tmp->level = level;

		p->rcpts = tmp;
	}
	while ((ptr = strtok(NULL, ",/:")) != NULL);
}


#ifndef CLIENT_ONLY
/*
** expand_recipients()
**
** This function expands a group into individual recipients.  This
** function is only called when using qpage in interactive mode.
**
**	Input:
**		p - the page containing the recipient list to expand
**
**	Returns:
**		nothing
*/
void
expand_recipients(PAGE *p)
{
	pgroup_t	*group;
	member_t	*member;
	rcpt_t		*list;
	rcpt_t		*tmp;


	list = p->rcpts;

	while (list) {
		if ((group = lookup(Groups, list->pager)) == NULL) {
			list = list->next;
			continue;
		}

		if (group->members) {
			free(list->pager);
			list->pager = strdup(group->members->pager->name);
			member = group->members->next;

			while (member) {
				tmp = (void *)malloc(sizeof(*tmp));
				(void)memset((char *)tmp, 0, sizeof(*tmp));
				tmp->pager = strdup(member->pager->name);
				tmp->next = list->next;
				list->next = tmp;
				list = tmp;

				member = member->next;
			}
		}

		list = list->next;
	}
}
#endif /* CLIENT_ONLY */


/*
** main()
*/
int
main(int argc, char **argv)
{
#ifndef CLIENT_ONLY
	struct sigaction	action;
	job_t			*joblist;
	short			port;
	char			msgid[100];
	int			sleeptime;
	int			showq;
#endif
	PAGE			p;
	time_t			t;
	time_t			holduntil;
	char			*coverage;
	char			*server;
	int			needpager;
	int			level;
	int			readmail;
	int			daemon;
	int			c;
	int			i;

	extern int		optind;
	extern char		*optarg;


	if (isatty(fileno(stdin)) == 0)
		Silent = TRUE;

#ifndef CLIENT_ONLY
	sleeptime = 0;
	port = 0;

	(void)memset((char *)&action, 0, sizeof(action));

#ifdef SA_NODEFER
	action.sa_flags |= SA_NODEFER;
#endif
#ifdef SA_NOCLDSTOP
	action.sa_flags |= SA_NOCLDSTOP;
#endif

	/*
	** Use sigaction() instead of signal() because we don't
	** want the SA_RESETHAND flag set.  That is, we want the
	** signal handler to remain in effect after the signal
	** is caught.
	*/
	action.sa_handler = (RETSIGTYPE(*)())sigalrm;
	if (sigaction(SIGALRM, &action, NULL) < 0)
		perror("sigaction(SIGALRM)");

	action.sa_handler = (RETSIGTYPE(*)())sigterm;
	if (sigaction(SIGTERM, &action, NULL) < 0)
		perror("sigaction(SIGTERM)");

	action.sa_handler = (RETSIGTYPE(*)())sigchld;
	if (sigaction(SIGCHLD, &action, NULL) < 0)
		perror("sigaction(SIGCHLD)");

	action.sa_handler = (RETSIGTYPE(*)())sighup;
	if (sigaction(SIGHUP, &action, NULL) < 0)
		perror("sigaction(SIGHUP)");

	action.sa_handler = (RETSIGTYPE(*)())sigusr1;
	if (sigaction(SIGUSR1, &action, NULL) < 0)
		perror("sigaction(SIGUSR1)");

	openlog(PROGRAM_NAME, LOG_PID|LOG_NDELAY|LOG_NOWAIT, SYSLOG_FACILITY);
#endif /* CLIENT_ONLY */

	server = NULL;
	coverage = NULL;
	holduntil = 0;
	level = DEFAULT_LEVEL;
	readmail = FALSE;
	daemon = FALSE;
	needpager = FALSE;
#ifndef CLIENT_ONLY
	showq = FALSE;
#endif

	(void)memset((char *)&p, 0, sizeof(p));

	while ((c = getopt(argc, argv, "a:c:C:df:hil:mp:P:q:Qs:t:v")) != -1) {
		switch (c) {
			case 'a':
			case 't':
				if ((t = parse_time(optarg)) == INVALID_TIME) {
					fprintf(stderr,
						"Invalid time: %s\n", optarg);
					return(-1);
				}

				holduntil = t;
				needpager = TRUE;
				break;

			case 'c':
				my_free(coverage);
				coverage = strdup(optarg);
				needpager = TRUE;
				break;

			case 'C':
				ConfigFile = strdup(optarg);
				break;

			case 'd':
				Debug++;
				break;

			case 'f':
				my_free(p.from);
				p.from = strdup(optarg);
				break;

			case 'h':
			case '?':
				do_usage(argv[0]);
				return(-1);

			case 'i':
				if (daemon) {
					fprintf(stderr, "Error: -i not valid with -q option\n");
					return(-1);
				}
#ifndef CLIENT_ONLY
				Interactive = TRUE;
				break;
#else
				fprintf(stderr, "Interactive mode not allowed.\n");
				return(-1);
#endif /* CLIENT_ONLY */

			case 'l':
				i = atoi(optarg);
				if ((i < 0) || (i > 11)) {
					fprintf(stderr, "Level must be between 0 and 11\n");
					return(-1);
				}

				level = i;
				needpager = TRUE;
				break;

			case 'm':
				readmail = TRUE;
				break;

			case 'p':
			case 'P':
				add_recipients(&p, optarg,
					coverage, holduntil, level);

				if (c == 'p') {
					my_free(coverage);
					coverage = NULL;
					holduntil = 0;
					level = DEFAULT_LEVEL;
				}
				needpager = FALSE;
				break;

			case 'q':
#ifndef CLIENT_ONLY
				if (Interactive) {
					fprintf(stderr, "Error: -q not valid with -i option\n");
					return(-1);
				}

				daemon = TRUE;
				sleeptime = atoi(optarg);
				break;
#else
				fprintf(stderr, "Server mode not allowed.\n");
				return(-1);
#endif /* CLIENT_ONLY */

			case 'Q':
#ifndef CLIENT_ONLY
				showq = TRUE;
				break;
#else
				fprintf(stderr, "Error: -Q support not compiled in.\n");
				return(-1);
#endif /* CLIENT_ONLY */

			case 's':
				server = optarg;
				break;

			case 'v':
				do_version();
				return(0);

			default:
				fprintf(stderr, "Internal program error\n");
				return(-1);
		}
	}

	if (ConfigFile != NULL) {
		if (daemon == FALSE && Interactive == FALSE) {
			fprintf(stderr,
				"Error: -C not allowed in client mode\n");

			return(-1);
		}
	}
	else
		ConfigFile = strdup(QPAGE_CONFIG);

#ifndef CLIENT_ONLY
	if (showq) {
		drop_root_privileges();

		if (get_qpage_config(ConfigFile) != 0) {
			fprintf(stderr, "Error reading configuration file\n");
			return(-1);
		}

		i = showqueue();
		printf("\n");
		return(i);
	}

	if (daemon) {
		if (sleeptime == 0) {
			drop_root_privileges();

			if (get_qpage_config(ConfigFile) != 0) {
				fprintf(stderr,
					"Error reading configuration file\n");
				return(-1);
			}

			(void)runqueue();
			return(0);
		}

		/*
		** Use a port specified on the command line.  This is
		** undocumented and unsupported.  Use at your own risk.
		*/
		if (optind < argc)
			port = atoi(argv[optind]);

		(void)become_daemon(sleeptime, port);

		/* NOT REACHED */
		return(-1);
	}
#endif /* CLIENT_ONLY */

	if (p.rcpts == NULL && optind < argc) {
		add_recipients(&p, argv[optind++], coverage, holduntil, level);
		needpager = FALSE;
	}

	if (p.rcpts == NULL) {
		fprintf(stderr, "No recipients!\n");
		return(-1);
	}

	if (needpager) {
		fprintf(stderr, "The -a, -c and -l options must come *before* the pagerids they affect\n");
		return(-1);
	}


	if (readmail)
		read_mail(&p);
	else
		p.message = command_line_message(argc-optind, &argv[optind]);

	if (p.message == NULL) {
		fprintf(stderr, "No message!\n");
		return(-1);
	}

	/*
	** fix the "from" address if necessary
	*/
	if (p.from == NULL)
		p.from = strdup(get_user());

	if (p.from[0] == '\0') {
		free(p.from);
		p.from = NULL;
	}


#ifndef CLIENT_ONLY
	if (Interactive) {
		/*
		** Previous versions let people use the -i
		** flag as root.  This resulted in lots of
		** confusion about why qpage worked fine
		** with the -i flag but not in regular
		** daemon mode.  I think it's better to
		** just make them both work the same.
		*/
		drop_root_privileges();

		if (get_qpage_config(ConfigFile) != 0) {
			fprintf(stderr, "Error reading configuration file\n");
			return(-1);
		}

		expand_recipients(&p);

		strip(&p.message);

		newmsgid(msgid);
		(void)strcat(msgid, "I");

		p.messageid = strdup(msgid);
		joblist = NULL;
		(void)insert_jobs(&joblist, &p);
		send_pages(joblist);
		i = 0;
	}
	else {
#endif /* CLIENT_ONLY */
		if (server == NULL)
			server = getenv("SNPP_SERVER");

#ifdef SNPP_SERVER_FILE
		if (server == NULL) {
			struct stat	statbuf;
			int		fd;


			if ((fd = open(SNPP_SERVER_FILE, O_RDONLY, 0)) >= 0) {
				if (fstat(fd, &statbuf) < 0) {
					perror("fstat() failed");
					return(-1);
				}

				i = statbuf.st_size;
				server = (void *)malloc(i+1);

				if (read(fd, server, i) != i) {
					perror("read() failed");
					return(-1);
				}

				server[i] = '\0';

				(void)close(fd);
			}
		}
#endif

		if (server == NULL)
			server = SNPP_SERVER;

		i = submit_page(&p, server);
#ifndef CLIENT_ONLY
	}

	closelog();
#endif
	return(i);
}
